### PR TITLE
refactor(server): stacks

### PR DESCRIPTION
**Pull request by jrasm91**
Refactor stacks to use a dedicated controller and set of endpoints.

TODO:
- Mobile
- e2e tests

### CODE REVIEW

File: server/src/repositories/stack.repository.ts [hunk 1]
```diff
@@ -1,23 +1,95 @@
...
+      },
+      relations: {
+        assets: {
+          exifInfo: true,
+        },
+      },
+    });
+  }
+
+  async create(entity: { ownerId: string; assetIds: string[] }): Promise<StackEntity> {
```
Comment (danieldietzler) [inline server/src/repositories/stack.repository.ts:31]: Do you think it's performance-wise fine to just do that in (well-readable) TS and not on the DB directly? Presumably you can assume that stacks are always "small", so it's even less of an issue?

File: server/src/services/stack.service.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    const stacks = await this.stackRepository.search({
+      ownerId: auth.user.id,
+      primaryAssetId: dto.primaryAssetId,
+    });
+
+    return stacks.map((stack) => mapStack(stack, { auth }));
+  }
+
+  async create(auth: AuthDto, dto: StackCreateDto): Promise<StackResponseDto> {
+    await this.access.requirePermission(auth, Permission.ASSET_READ, dto.assetIds);
```
Comment (danieldietzler) [inline server/src/services/stack.service.ts:32]: I'm not sure if read permissions should be sufficient here. In my mind `ASSET_UPDATE` would make more sense

File: server/src/services/stack.service.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  async get(auth: AuthDto, id: string): Promise<StackResponseDto> {
+    await this.access.requirePermission(auth, Permission.STACK_READ, id);
+    const stack = await this.findOrFail(id);
+    return mapStack(stack, { auth });
+  }
+
+  async update(auth: AuthDto, id: string, dto: StackUpdateDto): Promise<StackResponseDto> {
+    await this.access.requirePermission(auth, Permission.STACK_WRITE, id);
```
Comment (danieldietzler) [inline server/src/services/stack.service.ts:48]: Assuming you add new assets to the stack, you'd probably also want to enforce the same permissions for them you'd enforce when creating a stack

File: web/src/lib/utils/asset-utils.ts [hunk 4]
```diff
@@ -329,77 +332,52 @@ export const stackAssets = async (assets: AssetResponseDto[]) => {
...
-        stackParentId: parent.id,
+    const stack = await createStack({ stackCreateDto: { assetIds: assets.map(({ id }) => id) } });
+    notificationController.show({
+      message: $t('stacked_assets_count', { values: { count: stack.assets.length } }),
+      type: NotificationType.Info,
+      button: {
+        text: $t('view_stack'),
+        onClick() {
+          return assetViewingStore.setAssetId(stack.primaryAssetId);
+        },
```
Comment (danieldietzler) [inline web/src/lib/utils/asset-utils.ts:346]: I was not aware this is valid syntax lmao

**Review by danieldietzler — COMMENTED** [review]
Looking great so far!

File: server/src/dtos/stack.dto.ts [hunk 5]
```diff
@@ -1,9 +1,38 @@
...
 }
+
+export class StackResponseDto {
+  id!: string;
+  primaryAssetId!: string;
+  assets!: AssetResponseDto[];
+}
+
+export const mapStack = (stack: StackEntity, { auth }: { auth?: AuthDto }) => {
+  const primary = stack.assets.filter((asset) => asset.id === stack.primaryAssetId);
```
[the remark this answers is not in the document]
Comment (jrasm91) [reply server/src/dtos/stack.dto.ts:30]: This just made it easier to combine down below if it was always a list.

File: server/src/services/stack.service.ts (hunk 2 above)
Comment (jrasm91) [reply server/src/services/stack.service.ts:32]: Yeah, good point.

File: server/src/repositories/stack.repository.ts (hunk 1 above)
Comment (jrasm91) [reply server/src/repositories/stack.repository.ts:31]: The issue is we need to delete the stacks before creating new ones. I need the list of children from the stacks that are deleted. It seems a lot better to do that in a transaction so we don't delete them and then it fails and now the old stacks are gone.

File: server/src/services/stack.service.ts (hunk 3 above)
Comment (jrasm91) [reply server/src/services/stack.service.ts:48]: I believe update only supports changing the primary asset id, which already has to be in the stack, so I don't think any checks are necessary about asset visibility.

File: server/src/repositories/stack.repository.ts (hunk 1 above)
Comment (danieldietzler) [reply server/src/repositories/stack.repository.ts:31]: That makes a lot of sense, fair. Making it a fair bit more resilient

Comment (zackpollard) [reply server/src/repositories/stack.repository.ts:31]: Yea, this was what we decided in our initial talks, this all needs to happen in a transaction 😃

**Review by alextran1502 — APPROVED** [review]

**Comment by akostadinov** [comment]
Could this cause issues on android 1.114.0 https://github.com/immich-app/immich/issues/7286#issuecomment-2344562996 ?

**Pull request merged**